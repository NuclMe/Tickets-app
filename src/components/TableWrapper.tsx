import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import type { FC } from 'react';

import { device, fontSizes, colors } from '../styledVars';
import type { ISegment } from '../types';
import { getFlightTimeRange } from '../utils';

export const TableWrapper: FC<{ segments: ISegment }> = ({ segments }) => (
  <Table
    sx={{
      [`& .${tableCellClasses.root}`]: {
        borderBottom: 'none',
      },
    }}
  >
    <TableHead>
      <TableRow
        key={segments.duration}
        sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
      >
        <TableCell
          align="left"
          sx={{
            color: colors.grey,
            [`${device.tablet}`]: {
              fontSize: fontSizes.xs,
            },
          }}
        >
          {segments.origin} - {segments.destination}
        </TableCell>
        <TableCell
          align="center"
          sx={{
            color: colors.grey,
            [`${device.tablet}`]: {
              fontSize: fontSizes.xs,
            },
          }}
        >
          В ПУТИ
        </TableCell>
        <TableCell
          align="right"
          sx={{
            color: colors.grey,
            [`${device.tablet}`]: {
              fontSize: fontSizes.xs,
            },
          }}
        >
          {segments.stops.length} ПЕРЕСАДКИ{' '}
        </TableCell>
      </TableRow>
    </TableHead>
    <TableBody sx={{ borderBottom: 'none' }}>
      <TableRow>
        <TableCell
          align="left"
          sx={{
            [`${device.tablet}`]: {
              fontSize: fontSizes.xs,
            },
          }}
        >
          {getFlightTimeRange(segments.date, segments.duration)}
        </TableCell>
        <TableCell
          align="center"
          sx={{
            [`${device.tablet}`]: {
              fontSize: fontSizes.xs,
            },
          }}
        >
          {Math.floor(segments.duration / 60)}ч {segments.duration % 60}м
        </TableCell>
        <TableCell
          align="right"
          sx={{
            [`${device.tablet}`]: {
              fontSize: fontSizes.xs,
            },
          }}
        >
          {segments.stops.map((stop, idx, arr) => (
            <span key={stop}>
              {stop}
              {idx < arr.length - 1 ? ', ' : ''}
            </span>
          ))}
        </TableCell>
      </TableRow>
    </TableBody>
  </Table>
);
