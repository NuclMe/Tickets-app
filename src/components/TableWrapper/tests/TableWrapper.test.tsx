import { render, screen } from '@testing-library/react';

import { TableWrapper } from '../TableWrapper';

import {
  mockSegmentNoStops,
  mockSegmentOneStop,
  mockSegmentWithStops,
} from './mockData';

describe('TableWrapper', () => {
  it('renders the TableWrapper component via snapshot', () => {
    const { container } = render(
      <TableWrapper segments={mockSegmentWithStops} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render flight route information', () => {
    render(<TableWrapper segments={mockSegmentWithStops} />);

    expect(screen.getByText('MOW - HKT')).toBeInTheDocument();
    expect(screen.getByText('В ПУТИ')).toBeInTheDocument();
  });

  it('should display correct transfer count based on stops length', () => {
    render(<TableWrapper segments={mockSegmentWithStops} />);

    expect(screen.getByText('3 ПЕРЕСАДКИ')).toBeInTheDocument();
  });

  it('should format flight duration correctly', () => {
    render(<TableWrapper segments={mockSegmentWithStops} />);

    expect(screen.getByText('8ч 20м')).toBeInTheDocument();
  });

  it('should handle segment with no stops', () => {
    render(<TableWrapper segments={mockSegmentNoStops} />);

    expect(screen.getByText('0 ПЕРЕСАДКИ')).toBeInTheDocument();
    expect(screen.queryByText('LED')).not.toBeInTheDocument();
  });

  it('should handle segment with single stop (no comma)', () => {
    render(<TableWrapper segments={mockSegmentOneStop} />);

    expect(screen.getByText('1 ПЕРЕСАДКИ')).toBeInTheDocument();
    expect(screen.getByText('LED')).toBeInTheDocument();
    expect(screen.queryByText(/LED,/)).not.toBeInTheDocument();
  });

  it('should render flight time range', () => {
    render(<TableWrapper segments={mockSegmentWithStops} />);

    expect(screen.getByText(/\d{2}:\d{2} - \d{2}:\d{2}/)).toBeInTheDocument();
  });
});
