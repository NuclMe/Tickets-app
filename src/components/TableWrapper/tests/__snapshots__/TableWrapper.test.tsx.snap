// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableWrapper > renders the TableWrapper component via snapshot 1`] = `
<table
  class="MuiTable-root css-4o2jt9-MuiTable-root"
>
  <thead
    class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
  >
    <tr
      class="MuiTableRow-root MuiTableRow-head css-6phqnb-MuiTableRow-root"
    >
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-11trb53-MuiTableCell-root"
        scope="col"
      >
        MOW
         - 
        HKT
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-kl4v6g-MuiTableCell-root"
        scope="col"
      >
        В ПУТИ
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1cnmnjk-MuiTableCell-root"
        scope="col"
      >
        3
         ПЕРЕСАДКИ
         
      </th>
    </tr>
  </thead>
  <tbody
    class="MuiTableBody-root css-1o8qua4-MuiTableBody-root"
  >
    <tr
      class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeMedium css-tl43us-MuiTableCell-root"
      >
        05:00 - 13:20
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-ubhovo-MuiTableCell-root"
      >
        8
        ч 
        20
        м
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-qisdfz-MuiTableCell-root"
      >
        <span>
          LED
          , 
        </span>
        <span>
          AMS
          , 
        </span>
        <span>
          BCN
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
